register: add no-toolkit CLI option for pre-installed hosts

hardys opened the pull request:
Currently the sytem-agent-config is only generated as part of the cloudInitConfigs when installing, but when registering a pre-installed host it is useful to have the option to generate the system-agent-config without install.

This allows us for example to register a pre-installed SLEMicro host via a combustion script which installs then runs elemental-register without the `--install` flag, e.g I've been testing on a VM as below to prove the host registers and shows up in `MachineInventories` without and additional configuration

```
elemental-register --config-path /tmp/elemental-config.yaml --state-path=/tmp/elemental-state.yaml --emulate-tpm --install-local-agent-config --emulated-tpm-seed N
```

davidcassany commented:
@hardys thanks for the contribution. We will review the PR.

> This allows us for example to register a pre-installed SLEMicro host via a combustion script which installs then runs elemental-register without the `--install` flag, e.g I've been testing on a VM as below to prove the host registers and shows up in `MachineInventories` without and additional configuration

I have concerns about this procedure... I doubt you can do much more than actually register with this procedure. `/tmp` should be on tmpfs on SLEMicro, hence the configuration files are not persisting a reboot, hence due to the emulated TPM there might issues on reboot as the machine is likely  to re-register as a new machine with new random seed. Also upgrades managed by the UI are not going to work unless you use elemental images in the host.

In any case, registering a pre-installed host is something to think about so I believe the feature you aim to provide is valid. We'll gonna have a close look, thanks!

hardys commented:
> I have concerns about this procedure... I doubt you can do much more than actually register with this procedure. `/tmp` should be on tmpfs on SLEMicro, hence the configuration files are not persisting a reboot, hence due to the emulated TPM there might issues on reboot as the machine is likely to re-register as a new machine with new random seed. Also upgrades managed by the UI are not going to work unless you use elemental images in the host.
> 
> In any case, registering a pre-installed host is something to think about so I believe the feature you aim to provide is valid. We'll gonna have a close look, thanks!

Hi @davidcassany thanks for the feedback!

To clarify the example is only using `/tmp/ for developer testing (copy/paste from my VM env which it's true isn't a great example), the expected use of this will be via combustion in a transactional-update shell, so the actual usage would be like:

```
elemental-register --config-path /etc/elemental/config.yaml --state-path /etc/elemental/state.yaml --install-local-agent-config
```

And the new flag allows us to write out these two files which would also persist:

```
I0906 15:58:38.907820    2649 log.go:42] connection info file '/var/lib/elemental/agent/elemental_connection.json' written.
I0906 15:58:38.908297    2649 log.go:42] agent config file '/etc/rancher/elemental/agent/config.yaml' written.
```

The intention here is not to use the elemental upgrade path, but only to enable registration and integration into Rancher for vanilla SLEMIcro hosts (which is currently possible but you need some workarounds because these files are only generated on `--install`, I'm trying to make it cleaner with this PR)

@rdoxenham and @agracey may have further feedback as they've previously tested and proven the workaround-based approach mentioned above.

frelon commented:
@hardys just FYI, you have to run `make generate` in your branch and commit the changes for the Verify action to succeed :+1:

rdoxenham commented:
Thanks @hardys. Yes, we will not be using the Elemental UI based node upgrades here, the only requirement is to register the pre-installed SLE Micro hosts via the Elemental Toolkit. Node upgrades will be controlled/managed externally to Elemental. I'm not sure if this means that we'll need to add some additional logic to any components running on the host to check if it's actually running Elemental Teal or not, and to fail gracefully if it discovers that it's not, @davidcassany?

anmazzotti commented:
Thank you for the PR @hardys 

There are a few coding issues I think.
- The `Verify` Github action is failing. A `make generate` is needed.
- There is no test coverage.

About the functionality, I feel like you are bypassing the installation phase. 
This works at the moment, and we are trying to decouple the `elemental-toolkit` from `elemental-register` as much as possible, but still there is some coupling. For example the `elemental-system-agent` is used for the [Machine Reset](https://elemental.docs.rancher.com/next/reset) feature that heavily depends on the `elemental-toolkit` and on an Elemental installation that includes an Elemental recovery partition.
This feature is opt-in currently, but my point is that this may be difficult to maintain and we may break this new option eventually.

I would recommend a different approach, maybe a bit more hackish but probably better for you.
You can see how the `elemental-operator` generates this config [here](https://github.com/rancher/elemental-operator/blob/main/pkg/server/api_registration.go#L167).

Assuming you have access to the Rancher server, you should be able to extract the `URL`, `Token` and `Namespace` (this is the Registration namespace) alone.
The only information missing is the MachineInventory name, that matches the agent Secret name.  

So maybe an alternative approach is to add this information (the MachineInventory name) to the `elemental-register` [state](https://github.com/rancher/elemental-operator/blob/main/pkg/register/state.go#L32). You should then be able to run `elemental-register` for the first time to register, and then patch your agent config template with the correct `SecretName`, reading the state file. 

It is a bit hackish I reckon, but it's the least impactful approach.

Last but not least, I still have doubts on the use case. 
The `elemental-system-agent` currently is used for only 2 purposes:
- Bootstrap the machine with `k3s` or `rke2` components
- Reset the machine 

Are you going to use Rancher to bootstrap the machines then, or are they going to be pre-configured in that regard using combustion as well?

rdoxenham commented:
@anmazzotti - the idea is that we'll take vanilla SLE Micro hosts that may have already received partial configuration via ignition/combustion and execute `elemental-register` (which may actually happen during combustion) so that the machines can come under the command & control of Elemental. Primary reason for this will be to drive k3s/rke2 cluster deployment via existing Rancher methods; this is the primary use-case here, to re-use the Elemental onboarding and cluster management flow, but with vanilla SLE Micro. As already mentioned, we don't expect to be able to use the Elemental OS upgrade components with this approach.

hardys commented:
> Thank you for the PR @hardys
> 
> There are a few coding issues I think.
> 
> * The `Verify` Github action is failing. A `make generate` is needed.
> * There is no test coverage.

Verify issue is fixed (thanks for the pointer @frelon) - I can certainly add tests but would like to verify if we have consensus to accept the patch before spending time on that.

> About the functionality, I feel like you are bypassing the installation phase. This works at the moment, and we are trying to decouple the `elemental-toolkit` from `elemental-register` as much as possible, but still there is some coupling. For example the `elemental-system-agent` is used for the [Machine Reset](https://elemental.docs.rancher.com/next/reset) feature that heavily depends on the `elemental-toolkit` and on an Elemental installation that includes an Elemental recovery partition. This feature is opt-in currently, but my point is that this may be difficult to maintain and we may break this new option eventually.

Indeed, the point of the PR is to bypass the installation phase (which is already the default behaviour given that `--install` [defaults to false](https://github.com/rancher/elemental-operator/blob/main/cmd/register/main.go#L158)) - however currently that is inconvenient because the files exposed here are only generated when `--install` is enabled - the approach here makes life much more convenient at the point of registration for the pre-installed OS use-case.

Perhaps we can dig a little deeper into `we may break this new option eventually` - are you saying that registering pre-installed hosts (which appears to be the default as mentioned above) is not a valid use-case for Elemental?  @davidcassany appeared to indicate the opposite above so I'd like to clarify the position on that. 

> I would recommend a different approach, maybe a bit more hackish but probably better for you. You can see how the `elemental-operator` generates this config [here](https://github.com/rancher/elemental-operator/blob/main/pkg/server/api_registration.go#L167).
> 
> Assuming you have access to the Rancher server, you should be able to extract the `URL`, `Token` and `Namespace` (this is the Registration namespace) alone. The only information missing is the MachineInventory name, that matches the agent Secret name.
> So maybe an alternative approach is to add this information (the MachineInventory name) to the `elemental-register` [state](https://github.com/rancher/elemental-operator/blob/main/pkg/register/state.go#L32). You should then be able to run `elemental-register` for the first time to register, and then patch your agent config template with the correct `SecretName`, reading the state file.
> 
> It is a bit hackish I reckon, but it's the least impactful approach.

From my perspective this seems like a lot more work for a worse outcome, given that elemental-register already generates these files.  We'll just end up writing something very similar to the existing code in another tool - can you clarify why we can't just have elemental-register write out these files (given that the logic to do so already exists, it's just not enabled in the `--install=false` flow atm)?

> Last but not least, I still have doubts on the use case. The `elemental-system-agent` currently is used for only 2 purposes:
> 
> * Bootstrap the machine with `k3s` or `rke2` components
> * Reset the machine
> 
> Are you going to use Rancher to bootstrap the machines then, or are they going to be pre-configured in that regard using combustion as well?

As discussed above, the use-case here is just to use the Elemental onboarding and inventory flow, so we can boot SLEMicro hosts, have them show up in the elemental inventory then create clusters with the machines. We will not consume or depend on any of the day-2 features which are coupled to the specific base OS.

fgiudici reviewed (changes requested):
Hey @hardys , many thanks for this PR!
Overall the patch looks fine to me. Basically it introduces the option to do an initial registration without installing the Elemental OS (so, register and write the system agent config).
There are some requirements on the registering OS, i.e., it is on the OS provisioning task to provide and start an instance on the rancher-system-agent to leverage the system agent config written by the registration process.
Introducing this kind of scenario looks fine but I think there are few things to review before merging this patch:
1) we should pick carefully the elemental-register command option: we don't want to remove/change it later, so I would consider changing `install-local-agent-config` to something more generic referring to register enabling provisioning but not OS management (more future proof). Something like `noos-install` ... not that great, I know 😆  @hardys , @davidcassany , @anmazzotti , @frelon  any good name in mind?
2) we should label or annotate the MachineInventories tracking hosts which do not support Elemental OS management. While at this point I think we may either pass the registration type (install, reset, noos)  during the registration process or manage the "noos" option in the MachineRegistration, to keep it easy for this PR I think would be easier to pass the noos flag to the registration function and just pass one more annotation in [the sendAnnotation function](https://github.com/rancher/elemental-operator/blob/main/pkg/register/register.go#L275).
3) adding a unit-test would be nice.

hardys commented on cmd/register/main.go:
```diff
@@ -156,6 +168,7 @@ func newCommand(fs vfs.FS, client register.Client, stateHandler register.StateHa
 	_ = viper.BindPFlag("version", cmd.PersistentFlags().Lookup("version"))
 	cmd.Flags().BoolVar(&reset, "reset", false, "Reset the machine to its original post-installation state")
 	cmd.Flags().BoolVar(&installation, "install", false, "Install a new machine")
+	cmd.Flags().BoolVar(&installLocalAgentConfig, "install-local-agent-config", false, "Install agent config files to local filesystem")
```
Certainly open to suggestions on this - although one observation is the `--install` flag already defaults to false, so adding a new similar `--no-os-install` option could be confusing as to me they imply the same behaviour?

An alternative could be to unconditionally write out these files - currently they get written out to `/tmp` from [install.go](https://github.com/rancher/elemental-operator/blob/main/pkg/install/install.go#L53) but only when the `--install` flag is specified, we could write them unconditionally and add flags to override the default `/tmp` path to somewhere persistent?

hardys replied:
Now renamed to align with the suggested `--install --no-toolkit` approach

anmazzotti commented:
@fgiudici @hardys I propose to create a `--no-toolkit` flag. 
This flag should be **always on** when running `elemental-register` on an elemental-toolkitless host. 

This way we can also check it when `--install` or `--reset` have been issued. 

If `--install` and `--no-toolkit` are both on, then the `elemental-register` can just register and setup the `elemental-system-agent`  (this is the new feature added by this PR)

If `--install` and `--no-toolkit` are both on, then the `elemental-register` will do nothing. 

I also agree that adding an annotation is a great idea, so that from the operator side we can make some choices, for example suppressing the reset plan. This can be done here: https://github.com/rancher/elemental-operator/blob/main/controllers/machineinventory_controller.go#L190, by simply adding an additional check for this new annotation.

hardys commented:
> @fgiudici @hardys I propose to create a `--no-toolkit` flag. This flag should be **always on** when running `elemental-register` on an elemental-toolkitless host.
> 
> This way we can also check it when `--install` or `--reset` have been issued.

This sounds reasonable to me, but one clarification - does `no-toolkit` only imply skipping the day-2 OS management operations or also cluster bootstrapping?  (As previously discussed we do want to leverage elemental-system-agent to kick off the cluster bootstrapping via Rancher) 

> I also agree that adding an annotation is a great idea, so that from the operator side we can make some choices, for example suppressing the reset plan. This can be done here: https://github.com/rancher/elemental-operator/blob/main/controllers/machineinventory_controller.go#L190, by simply adding an additional check for this new annotation.

That sounds good - I'll raise an issue as previously discussed so we can enumerate the tasks beyond this initial change - handling such annotation logic sounds like a follow-up task/PR to me?

anmazzotti commented:
Cluster bootstrapping is handled by the `elemental-system-agent`, so the `--no-toolkit` will not have influence on that. 

Regarding the follow up, I would rather have the new annotation and the reset plan suppression included in this PR. 
I think the reset is the only thing that has chances to break with this new feature. Since reset is an opt-in feature it's unlikely to happen, but since the change is rather small I would add it here for safety.

However I don't have a strong opinion on this. It can be a follow up as well. 

We should also add some documentation regarding the new annotation and argument in the `elemental-docs`, so having an issue to track all PRs is great.

anmazzotti commented:
Also sorry for that, I just noticed I made a mistake, I previously meant: 

If `--reset` and `--no-toolkit` are both on, then the `elemental-register` will do nothing.

hardys commented on cmd/register/main.go:
```diff
@@ -81,6 +82,10 @@ func newCommand(fs vfs.FS, client register.Client, stateHandler register.StateHa
 				return nil
 			}
+			// Validate installation options
+			if installation && installLocalAgentConfig {
```
This logic is removed since reworking to align with the suggested `--install --no-toolkit` approach

hardys commented:
@anmazzotti @fgiudici thanks for the detailed review feedback!

I have now reworked the CLI part to align with the suggested `--install --no-toolkit` approach, and also added a new `os.unmanaged` annotation as suggested (open to other suggestions re naming)

If you could please re-review to confirm if this aligns with your suggestions that would be great, provided I have interpreted them correctly I will then add tests/docs so we can hopefully iterate towards merging.

fgiudici commented on api/v1beta1/machineinventory_types.go:
```diff
@@ -29,6 +29,7 @@ var (
 	PlanTypeBootstrap                                      = "bootstrap"
 	PlanTypeReset                                          = "reset"
 	MachineInventoryResettableAnnotation                   = "elemental.cattle.io/resettable"
```
gofmt complaining: the whole var section should have values moved by one space more (aligned from 'corev1.SecretType'). Quickly fixable with:
`gofmt -e -d -s -w api/v1beta1/machineinventory_types.go`

fgiudici reviewed (commented):
Hey @hardys , patch looks good to me, thanks!
Just one gofmt formatting and  the missing tests than I would merge 👍🏼 
for the docs it should land on rancher/elemental-docs as a separate PR.

hardys commented:
> Hey @hardys , patch looks good to me, thanks!
> Just one gofmt formatting and the missing tests than I would merge 👍🏼
> for the docs it should land on rancher/elemental-docs as a separate PR.

Thanks - I think I resolved the lint/generate issues and will work on adding some tests 👍

hardys commented:
@fgiudici I added some basic tests, as an additional commit because they depend on the API changes introduced in the second commit, let me know if you think this is sufficient and I'll work on a docs PR, thanks!

fgiudici reviewed (approved):
Thanks @hardys , awesome!

fgiudici commented:
The errors in the verify job are not related to this PR. Verify failure tracked in issue #522 .

fgiudici commented:
rebased the branch
